tractor/_portal.py

```diff
@@ -67,34 +53,69 @@ class Portal:
     """
     def __init__(self, channel):
         self.channel = channel
+        # when this is set to a tuple returned from ``_submit()`` then
+        # it is expected that ``result()`` will be awaited at some point
+        # during the portal's lifetime
         self._result = None
+        self._expect_result = None
+        self._errored = False
 
     async def aclose(self):
         log.debug(f"Closing {self}")
         # XXX: won't work until https://github.com/python-trio/trio/pull/460
         # gets in!
         await self.channel.aclose()
 
+    async def _submit(self, ns, func, **kwargs):
+        """Submit a function to be scheduled and run by actor, return the
+        associated caller id, response queue, response type str,
+        first message packet as a tuple.
+
+        This is an async call.
+        """
+        # ship a function call request to the remote actor
+        cid, q = await current_actor().send_cmd(self.channel, ns, func, kwargs)
+
+        # wait on first response msg and handle (this should be
+        # in an immediate response)
+        first_msg = await q.get()
+        functype = first_msg.get('functype')
+
+        if functype == 'function' or functype == 'asyncfunction':
+            resp_type = 'return'
+        elif functype == 'asyncgen':
+            resp_type = 'yield'
```

Comment:
Have you considered using an enum here instead of strings?

Reply:
Added #27 for this.